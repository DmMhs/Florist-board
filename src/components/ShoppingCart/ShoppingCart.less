.ShoppingCart {
  position: fixed;
  top: 0;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  width: 100%;
  height: 100%;
  background-color: rgba(51, 51, 51, 0.5);
  z-index: 9;
  .cart-content {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    position: relative;
    width: 75%;
    min-height: 200px;
    border-radius: 7px;
    height: auto;
    background-color: white;
    z-index: 4;
    .order {
      height: auto;
      display: flex;
      flex-direction: column;
      justify-content: flex-start;
      align-items: center;
      .order-btn {
        margin-bottom: 2rem;
        padding: 5px;
        background-color: rgb(137, 184, 137);
        font-size: 1.5rem;
        color: white;
        border: none;
        &:hover {
          cursor: pointer;
          background-color: rgb(110, 207, 110);
        }
      }
    }
    .close-cart {
      position: absolute;
      top: 20px;
      right: 25px;
      font-size: 1.3rem;
      color: #333333;
      z-index: 4;
      cursor: pointer;
      &:hover {
        color: tomato;
      }
    }
    p,
    h4 {
      font-weight: lighter;
      font-size: 2rem;
    }
    ul {
      list-style: none;
      padding: 0;
      hr {
        max-width: 50%;
        border: 0.5px solid #eee;
      }
      li {
        display: flex;
        padding: 7px;
        font-size: 1.3rem;
        width: 800px;
        font-weight: 300;
        list-style: none;
        border: 1px solid #ccc;
        box-sizing: border-box;
        margin: 5px 0;
        &:first-child {
          border: none;
        }

        span {
          padding: 9px;
          display: flex;
          flex-direction: row;
          justify-content: center;
          flex-wrap: wrap;
          flex-basis: 20%;
          align-items: center;
          color: #333333;
          &.product-name {
            justify-content: space-between;
          }
          i {
            cursor: pointer;
            margin: 0 3px;
            font-size: 1.3rem;
            color: #353434;
            &.minus:hover {
              color: tomato;
            }
            &.plus:hover {
              color: rgb(126, 194, 126);
            }
            &.info {
              cursor: pointer;
              font-size: 1.4rem;
              color: skyblue;
              &:hover {
                color: rgb(0, 183, 255);
              }
            }
            &.remove:hover {
              color: tomato;
            }
          }
        }
        .amount {
          display: flex;
          flex-direction: row;
          justify-content: space-between;
          align-items: center;
          border-left: 1px solid #eee;
          border-right: 1px solid #eee;
          box-sizing: border-box;
        }
      }
    }
  }
}

@media screen and (max-width: 478px) {
  .ShoppingCart {
    position: fixed;
    width: 100%;
    margin: 0 auto;
    .cart-content {
      display: block;
      width: 100vw;
      min-height: 100%;
      overflow-y: auto;
      border-radius: 0px;
      margin: 0 auto;
      ul {
        width: 100%;
        margin: 4rem 0;
        li {
          flex-direction: column;
          padding: 3px;
          font-size: 1rem;
          width: 260px;
          margin: 5px auto;
          border-radius: 7px;
          &:first-child {
            display: none;
          }
          span {
            position: relative;
            &.product-name {
              justify-content: center;
            }
            i {
              &.info {
                margin: 0 5px;
              }
            }
          }
          .amount {
            justify-content: center;
            i {
              margin: 0 10px;
            }
          }
        }
      }
      .order {
        p {
          font-size: 1.5rem;
        }

        .order-btn {
          font-size: 1rem;
          font-weight: lighter;
        }
      }
      .close-cart {
        top: 2rem;
        right: 2rem;
      }
    }
  }
}

@media screen and (min-width: 478.1px) and (max-width: 576px) {
  .ShoppingCart {
    position: fixed;
    .cart-content {
      display: block;
      width: 100%;
      min-height: 100%;
      overflow-y: auto;
      border-radius: 0px;
      ul {
        width: 100vw;
        margin-top: 4rem;
        li {
          flex-direction: column;
          padding: 3px;
          font-size: 1rem;
          width: 400px;
          margin: 5px auto;
          border-radius: 7px;
          &:first-child {
            display: none;
          }
          span {
            position: relative;
            &.product-name {
              justify-content: center;
            }
            i {
              &.info {
                margin: 0 5px;
              }
            }
          }
          .amount {
            justify-content: center;
            i {
              margin: 0 10px;
            }
          }
        }
      }
      .close-cart {
        top: 2rem;
        right: 2rem;
      }
    }
  }
}

@media screen and (min-width: 576.1px) and (max-width: 768px) {
  .ShoppingCart {
    position: fixed;
    .cart-content {
      display: block;
      width: 100vw;
      min-height: 100%;
      overflow-y: auto;
      border-radius: 0px;
      ul {
        width: 100%;
        li {
          padding: 3px;
          font-size: 1rem;
          width: 520px;
          margin: 5px auto;
          span {
            position: relative;
            i {
              &.info {
                position: absolute;
                top: 10px;
                right: 10px;
              }
            }
          }
        }
      }
    }
  }
}

@media screen and (min-width: 768.1px) and (max-width: 921px) {
  .ShoppingCart {
    position: fixed;
    .cart-content {
      display: block;
      width: 100%;
      min-height: 100%;
      overflow-y: auto;
      border-radius: 0px;
      ul {
        li {
          font-size: 1.1rem;
          width: 700px;
          margin: 5px auto;
          span {
            position: relative;
            i {
              &.info {
                position: absolute;
                top: 10px;
                right: 10px;
              }
            }
          }
        }
      }
    }
  }
}
@media screen and (min-width: 921.1px) and (max-width: 1200px) {
  .ShoppingCart {
    position: fixed;
    .cart-content {
      display: block;
      width: 100%;
      min-height: 100%;
      overflow-y: auto;
      border-radius: 0px;
      ul {
        li {
          font-size: 1.2rem;
          width: 800px;
          margin: 5px auto;
          span {
            position: relative;
            i {
              &.info {
                position: absolute;
                top: 10px;
                right: 10px;
              }
            }
          }
        }
      }
    }
  }
}
