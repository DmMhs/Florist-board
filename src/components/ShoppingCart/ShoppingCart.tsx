import React, { Component } from 'react';

import { CartItem } from '../../models/CartItem';
import './ShoppingCart.less';

interface ShoppingCartProps {
  cartItems: CartItem[];
  showCart: boolean;
  closeCart:
    | ((event: React.MouseEvent<HTMLElement, MouseEvent>) => void)
    | undefined;
  remove: any;
}
interface ShoppingCartState {
  cartProducts: Partial<CartItem> & { amount: number }[];
}

class ShoppingCart extends Component<ShoppingCartProps, ShoppingCartState> {
  static getDerivedStateFromProps(
    props: ShoppingCartProps,
    state: ShoppingCartState
  ) {
    let initialSum = 0;
    props.cartItems.map((i: CartItem) => {
      initialSum += +i.price!;
    });
    initialSum.toFixed(2);
    return {
      cartProducts: props.cartItems,
      totalSum: initialSum
    };
  }
  constructor(props: any) {
    super(props);
    this.state = {
      cartProducts: []
    };
  }

  reduceAmountClickedHandler = (amount: number, index: number) => {
    if (amount > 1) {
      const newCart = { ...this.state.cartProducts };
      newCart[index].amount = newCart[index].amount - 1;
      this.setState({
        cartProducts: newCart
      });
    }
  };
  increaseAmountClickedHandler = (amount: number, index: number) => {
    const newCart = { ...this.state.cartProducts };
    newCart[index].amount = newCart[index].amount + 1;
    this.setState({
      cartProducts: newCart
    });
  };
  removeClickedHandler = (index: number) => {
    const newCartProducts = this.state.cartProducts.splice(index, 1);
    this.setState({
      cartProducts: newCartProducts
    });
  };
  render() {
    const style = {
      display: this.props.showCart ? 'flex' : 'none'
    };
    let totalPrice = 0;
    const cartItems = this.state.cartProducts;
    const cartItemsList = cartItems.map(
      (i: Partial<CartItem> & { amount: number }, index: number) => {
        totalPrice += +i.price! * i.amount!;
        return (
          <li className="cart-item" key={index}>
            <div
              className="cart-item-img"
              style={{
                display: 'flex',
                flexBasis: '20%',
                background: `url(${i.images![0]})`,
                width: '120px',
                height: '120px',
                backgroundSize: 'cover',
                backgroundRepeat: 'no-repeat',
                backgroundPosition: 'center'
              }}
            />
            <span>{i.title}</span>
            <span className="amount">
              <i
                className="far fa-minus-square"
                onClick={() => this.reduceAmountClickedHandler(i.amount, index)}
              />
              {i.amount}
              <i
                className="far fa-plus-square"
                onClick={() =>
                  this.increaseAmountClickedHandler(i.amount, index)
                }
              />
            </span>
            <span>
              {(i.price! * i.amount!).toFixed(2)}
              {i.currency}
            </span>
            <span>
              <i
                className="fas fa-times"
                onClick={() => this.props.remove(index)}
              />
            </span>
          </li>
        );
      }
    );

    const cartContent =
      this.state.cartProducts.length > 0 ? (
        <div>
          <ul>
            <li>
              <span>PRODUCT</span>
              <span />
              <span>AMOUNT</span>
              <span>PRICE</span>
              <span />
            </li>
            <hr />
            {cartItemsList}
          </ul>
          <div className="order">
            <p>Total Sum: {totalPrice.toFixed(2)}$</p>
            <button type="button" className="order-btn">
              ORDER
            </button>
          </div>
        </div>
      ) : (
        <h4>
          cart is empty <i className="far fa-frown-open" />
        </h4>
      );
    return (
      <div className="ShoppingCart" style={style}>
        <div className="cart-content">
          <i
            className="fas fa-times close-cart"
            onClick={this.props.closeCart}
          />
          {cartContent}
        </div>
      </div>
    );
  }
}

export default ShoppingCart;
