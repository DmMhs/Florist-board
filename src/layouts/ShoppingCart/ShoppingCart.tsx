import React, { Component } from 'react';

import { CartItem } from '../../models/CartItem';
import './ShoppingCart.less';
import ShoppingCartItem from './ShoppingCartItem/ShoppingCartItem';
import ShoppingCartContent from './ShoppingCartContent/ShoppingCartContent';

interface ShoppingCartProps {
  cartItems: CartItem[];
  showCart: boolean;
  closeCart:
    | ((event: React.MouseEvent<HTMLElement, MouseEvent>) => void)
    | undefined;
  remove: ((index: number) => void) | undefined;
}
interface ShoppingCartState {
  cartProducts: Partial<CartItem> & { amount: number }[];
}

class ShoppingCart extends Component<ShoppingCartProps, ShoppingCartState> {
  public static getDerivedStateFromProps(
    props: ShoppingCartProps,
    state: ShoppingCartState
  ) {
    let initialSum = 0;
    props.cartItems.map((i: CartItem) => {
      initialSum += +i.price!;
    });
    initialSum.toFixed(2);
    return {
      cartProducts: props.cartItems,
      totalSum: initialSum
    };
  }
  constructor(props: ShoppingCartProps) {
    super(props);
    this.state = {
      cartProducts: []
    };
  }

  private reduceAmountClickedHandler = (amount: number, index: number) => {
    if (amount > 1) {
      const newCart = { ...this.state.cartProducts };
      newCart[index].amount = newCart[index].amount - 1;
      this.setState({
        cartProducts: newCart
      });
    }
  };
  private increaseAmountClickedHandler = (index: number) => {
    const newCart = { ...this.state.cartProducts };
    newCart[index].amount = newCart[index].amount + 1;
    this.setState({
      cartProducts: newCart
    });
  };
  private removeClickedHandler = (index: number) => {
    const newCartProducts = this.state.cartProducts.splice(index, 1);
    this.setState({
      cartProducts: newCartProducts
    });
  };
  public render() {
    const { showCart, closeCart, remove } = this.props;
    const { cartProducts } = this.state;

    const style = {
      display: showCart ? 'flex' : 'none'
    };

    let totalPrice = 0;
    let cartItemPrice = 0;
    const cartItems = cartProducts;

    const cartItemsList = cartItems.map(
      (i: Partial<CartItem> & { amount: number }, index: number) => {
        totalPrice += +i.price! * i.amount!;
        cartItemPrice = +(i.price! * i.amount!).toFixed(2);
        return (
          <ShoppingCartItem
            remove={remove!.bind(this, index)}
            increaseAmount={
              this.increaseAmountClickedHandler.bind(this, index) as
                | ((event: React.MouseEvent<HTMLElement, MouseEvent>) => void)
                | undefined
            }
            reduceAmount={
              this.reduceAmountClickedHandler.bind(this, i.amount, index) as
                | ((event: React.MouseEvent<HTMLElement, MouseEvent>) => void)
                | undefined
            }
            images={i.images as string[]}
            title={i.title as string}
            title_ua={i.title_ua as string}
            id={i.id as string}
            amount={i.amount}
            cartItemPrice={cartItemPrice}
            currency={i.currency as string}
            key={index}
          />
        );
      }
    );

    return (
      <div className="ShoppingCart" style={style}>
        <div className="cart-content">
          <i className="fas fa-times close-cart" onClick={closeCart} />
          <ShoppingCartContent
            cartItemsList={cartItemsList as JSX.Element[]}
            totalPrice={totalPrice}
            fixEmptyCart={closeCart}
          />
        </div>
      </div>
    );
  }
}

export default ShoppingCart;
