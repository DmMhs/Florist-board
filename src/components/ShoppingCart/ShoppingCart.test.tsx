import React from 'react';
import { shallow, mount } from 'enzyme';

import ShoppingCart from './ShoppingCart';
const cartItem = {
  title: 'qwerty',
  images: ['asd', 'qwr'],
  price: 100,
  amount: 1,
  currency: 'usd',
  available: true,
  key: 10,
  id: 'asgtqwr',
  inCart: false,
  remove: undefined
};
it('ShoppingCartItem matches a snapshot', () => {
  const wrapper = shallow(
    <ShoppingCart
      cartItems={[cartItem, cartItem]}
      showCart={true}
      closeCart={() => {}}
      remove={() => {}}
    />
  );
  expect(wrapper).toMatchSnapshot();
});

it("Increasing cart item's amount changes the state", () => {
  const wrapper = shallow(
    <ShoppingCart
      cartItems={[cartItem]}
      showCart={true}
      closeCart={() => {}}
      remove={() => {}}
    />
  );
  const instance = wrapper.instance();
  instance.increaseAmountClickedHandler(0);
  expect(instance.state.cartProducts[0].amount).toEqual(2);
});

it("Reducing cart item's amount changes the state", () => {
  const cartItem2 = {
    title: 'qwerty',
    images: ['asd', 'qwr'],
    price: 100,
    amount: 4,
    currency: 'usd',
    available: true,
    key: 10,
    id: 'asgtqwr',
    inCart: false,
    remove: undefined
  };
  const wrapper = shallow(
    <ShoppingCart
      cartItems={[cartItem2]}
      showCart={true}
      closeCart={() => {}}
      remove={() => {}}
    />
  );
  const instance = wrapper.instance();
  instance.reduceAmountClickedHandler(2, 0);
  expect(instance.state.cartProducts[0].amount).toEqual(3);
});
