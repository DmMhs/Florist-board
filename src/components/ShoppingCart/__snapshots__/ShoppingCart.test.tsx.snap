// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShoppingCart works as expected ShoppingCartItem matches a snapshot 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <withRouter(AppContextProvider)>
      <Route>
        <AppContextProvider
          history={
            Object {
              "action": "POP",
              "block": [Function],
              "createHref": [Function],
              "go": [Function],
              "goBack": [Function],
              "goForward": [Function],
              "length": 1,
              "listen": [Function],
              "location": Object {
                "hash": "",
                "pathname": "/",
                "search": "",
                "state": undefined,
              },
              "push": [Function],
              "replace": [Function],
            }
          }
          location={
            Object {
              "hash": "",
              "pathname": "/",
              "search": "",
              "state": undefined,
            }
          }
          match={
            Object {
              "isExact": true,
              "params": Object {},
              "path": "/",
              "url": "/",
            }
          }
        >
          <ShoppingCart
            cartItems={Array []}
            closeCart={[Function]}
            remove={[Function]}
            showCart={true}
          >
            <div
              className="ShoppingCart"
              style={
                Object {
                  "display": "flex",
                }
              }
            >
              <div
                className="cart-content"
              >
                <i
                  className="fas fa-times close-cart"
                  onClick={[Function]}
                />
                <ShoppingCartContent
                  cartItemsList={Array []}
                  fixEmptyCart={[Function]}
                  totalPrice={0}
                >
                  <div>
                    <h4>
                      cart is empty
                    </h4>
                    <button
                      className="fix"
                      onClick={[Function]}
                      type="button"
                    >
                      <span>
                        LET'S FIX IT
                      </span>
                       
                      <i
                        className="far fa-smile-wink"
                      />
                    </button>
                  </div>
                </ShoppingCartContent>
              </div>
            </div>
          </ShoppingCart>
        </AppContextProvider>
      </Route>
    </withRouter(AppContextProvider)>
  </Router>
</BrowserRouter>
`;
